File: backend-services/db-mcp-server/src/auth/resource-server.ts
import { createRemoteJWKSet, jwtVerify, type JWTPayload, type JWTVerifyGetKey } from 'jose';
import { McpError } from '../errors';

/**
 * Inbound resource-server authentication for the DB MCP server. Validates the
 * caller's audience-restricted `client_credentials` token (full signature,
 * issuer, audience) and pins the authorized party (`azp`) so ONLY the SQL
 * analyst agent — never a user token, never another service — can reach the
 * data tools. Default deny: any failure throws and the request is rejected.
 */
export interface VerifiedCaller {
  /** Authorized party (calling client id), already checked against the allowlist. */
  readonly azp: string;
  /** Token subject (the service account), for audit only. */
  readonly subject: string;
}

export interface ResourceServerConfig {
  readonly issuerUrl: string;
  readonly jwksUri: string;
  readonly audience: string;
  readonly authorizedParties: readonly string[];
}

export class ResourceServer {
  private readonly getKey: JWTVerifyGetKey;

  constructor(private readonly config: ResourceServerConfig) {
    this.getKey = createRemoteJWKSet(new URL(config.jwksUri));
  }

  async verify(authorizationHeader: string | undefined): Promise<VerifiedCaller> {
    const token = this.extractBearer(authorizationHeader);
    let payload: JWTPayload;
    try {
      ({ payload } = await jwtVerify(token, this.getKey, {
        issuer: this.config.issuerUrl,
        audience: this.config.audience,
        algorithms: ['RS256', 'ES256'],
      }));
    } catch {
      throw new McpError('unauthenticated', 'The provided token is invalid or expired.');
    }

    const azp = typeof payload['azp'] === 'string' ? payload['azp'] : undefined;
    if (!azp || !this.config.authorizedParties.includes(azp)) {
      throw new McpError('unauthenticated', 'The token was issued to an unauthorized party.');
    }
    const subject = typeof payload.sub === 'string' ? payload.sub : azp;
    return { azp, subject };
  }

  private extractBearer(header: string | undefined): string {
    if (!header || !header.startsWith('Bearer ')) {
      throw new McpError('unauthenticated', 'A bearer token is required.');
    }
    const token = header.slice('Bearer '.length).trim();
    if (!token) {
      throw new McpError('unauthenticated', 'A bearer token is required.');
    }
    return token;
  }
}

File: agents/at-sql-analyser/src/at_sql_analyser/auth/resource_server.py
"""Inbound resource-server authentication for the agent's A2A surface.

Validates the caller's audience-restricted token (signature via JWKS, issuer,
``aud: nova-agent-sql-analyst``, ``exp``) and pins the authorized party
(``azp``) to the Celery worker — the ONLY client allowed to task this agent.
Default deny: any failure raises and the request is rejected. Mirrors the
orchestrator gateway's PyJWT pattern, hardened with `azp` pinning.
"""

from __future__ import annotations

from dataclasses import dataclass

import jwt
from jwt import PyJWKClient

_ALLOWED_ALGS = ["RS256", "ES256"]


@dataclass(frozen=True)
class VerifiedCaller:
    azp: str
    subject: str


class AuthError(Exception):
    """Raised when an inbound token is missing, invalid, or unauthorized."""


class ResourceServer:
    def __init__(
        self,
        *,
        jwks_uri: str,
        issuer_url: str,
        audience: str,
        authorized_parties: tuple[str, ...],
    ) -> None:
        self._jwks_client = PyJWKClient(jwks_uri)
        self._issuer = issuer_url
        self._audience = audience
        self._authorized_parties = set(authorized_parties)

    def verify(self, authorization_header: str) -> VerifiedCaller:
        if not authorization_header.startswith("Bearer "):
            raise AuthError("bearer token required")
        token = authorization_header.removeprefix("Bearer ").strip()
        if not token:
            raise AuthError("bearer token required")
        try:
            signing_key = self._jwks_client.get_signing_key_from_jwt(token)
            claims = jwt.decode(
                token,
                signing_key.key,
                algorithms=_ALLOWED_ALGS,
                audience=self._audience,
                issuer=self._issuer,
                options={"require": ["exp", "iss", "aud"]},
            )
        except jwt.PyJWTError as exc:
            raise AuthError("invalid service token") from exc

        azp = claims.get("azp")
        if not isinstance(azp, str) or azp not in self._authorized_parties:
            raise AuthError("token issued to an unauthorized party")
        subject = claims.get("sub")
        return VerifiedCaller(azp=azp, subject=str(subject) if subject else azp)
